Fix functions not being kept for debugger

augusto2112 opened the pull request:
Explanation: At Onone, many types of functions (anything user written, compiler generated setters and getters, etc), should be kept in the final binary so they're accessible by the debugger. They were previously being optimized out.
Scope: changes which functions should be kept for the debugger or not.
Risk: Low, affects whether a function should be optimized out or not, but not the actual logic of IRGening functions.
Testing: Added a compiler test. Modified a few tests to match the current expected behavior.
Issue: rdar://126763340
Reviewer: @adrian-prantl @al45tair @eeckstein

augusto2112 committed:
Fix functions not being kept for debugger

At Onone, many types of functions (anything user written, compiler
generated setters and getters, etc), should be kept in the final
binary so they're accessible by the debugger.

rdar://126763340
(cherry picked from commit 529845056ba835dbcaed04fe0290a364e87e5e30)

augusto2112 commented:
@swift-ci test

augusto2112 commented:
https://github.com/apple/llvm-project/pull/8740
@swift-ci test

slavapestov commented on lib/SIL/IR/SILFunction.cpp:
```diff
@@ -16,6 +16,7 @@
 #include "swift/AST/Availability.h"
 #include "swift/AST/Expr.h"
 #include "swift/AST/GenericEnvironment.h"
+#include "swift/AST/IRGenOptions.h"
```
I don't think it makes sense for files in lib/SIL to depend on IRGen headers. Can you find another way to plumb this through?

augusto2112 replied:
Since this already merged in main I opened https://github.com/apple/swift/pull/73727 to address this issue

augusto2112 replied:
I cherry-picked https://github.com/apple/swift/pull/73727 here as well

adrian-prantl commented:
@swift-ci test windows

augusto2112 committed:
Remove dependency of SILFunction in IRGenOptions

SILFunction::shouldBePreservedForDebugger checks if some optimizations
are enabled to decide whether a function should be preserved so its
accessible form the debugger or not. Some of these settings used to live
only in IRGenOptions making SILFunction depend on IRGenOptions.

(cherry picked from commit 4aec4e7cc4d8bea8c297d59c90db2e374b0b7d05)

augusto2112 commented:
@swift-ci test
